src/vs/workbench/services/textMate/browser/workerHost/textMateWorkerTokenizerController.ts

```diff
@@ -0,0 +1,155 @@
+import { TextMateTokenizationWorker } from 'vs/workbench/services/textMate/browser/worker/textMate.worker';
+import type { StateDeltas } from 'vs/workbench/services/textMate/browser/workerHost/textMateWorkerHost';
+import { applyStateStackDiff, StateStack } from 'vscode-textmate';
+
+export class TextMateWorkerTokenizerController extends Disposable {
+	private _pendingChanges: IModelContentChangedEvent[] = [];
+
+	/**
+	 * These states will eventually equal the worker states.
+	 * _states[i] stores the state at the end of line number i+1.
+	 */
+	private readonly _states = new ContiguousGrowingArray<StateStack | null>(null);
+
+	constructor(
+		private readonly _model: ITextModel,
+		private readonly _worker: TextMateTokenizationWorker,
+		private readonly _languageIdCodec: ILanguageIdCodec,
+		private readonly _backgroundTokenizationStore: IBackgroundTokenizationStore,
+		private readonly _initialState: StateStack,
+	) {
+		super();
+
+		this._register(this._model.onDidChangeContent((e) => {
+			this._worker.acceptModelChanged(this._model.uri.toString(), e);
+			this._pendingChanges.push(e);
+		}));
+
+		this._register(this._model.onDidChangeLanguage((e) => {
+			const languageId = this._model.getLanguageId();
+			const encodedLanguageId =
+				this._languageIdCodec.encodeLanguageId(languageId);
+			this._worker.acceptModelLanguageChanged(
+				this._model.uri.toString(),
+				languageId,
+				encodedLanguageId
+			);
+		}));
+
+		const languageId = this._model.getLanguageId();
+		const encodedLanguageId = this._languageIdCodec.encodeLanguageId(languageId);
+		this._worker.acceptNewModel({
+			uri: this._model.uri,
+			versionId: this._model.getVersionId(),
+			lines: this._model.getLinesContent(),
+			EOL: this._model.getEOL(),
+			languageId,
+			encodedLanguageId,
+		});
+	}
+
+	public override dispose(): void {
+		super.dispose();
+		this._worker.acceptRemovedModel(this._model.uri.toString());
+	}
+
+	/**
+	 * This method is called from the worker through the worker host.
+	 */
+	public setTokensAndStates(versionId: number, rawTokens: ArrayBuffer, stateDeltas: StateDeltas[]): void {
+		// _states state, change{k}, ..., change{versionId}, state delta base, change{j}, ..., change{m}, current renderer state
+		//                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                    ^^^^^^^^^^^^^^^^^^^^^^^^^
+		//                | past changes                                       | future states
+
+		// Apply past changes to _states
+		while (
+			this._pendingChanges.length > 0 &&
+			this._pendingChanges[0].versionId <= versionId
+		) {
+			const change = this._pendingChanges.shift()!;
+			const op = lineArrayEditFromModelContentChange(change.changes);
+			op.applyTo(this._states);
+		}
+
+		const curToFutureTransformer1 = MonotonousIndexTransformer.fromMany(
+			this._pendingChanges.map((c) => lineArrayEditFromModelContentChange(c.changes))
+		);
+
```

Comment:
What's the difference between `curToFutureTransformer1` and `curToFutureTransformer` ?